File: src/main/java/me/lordsaad/modeoff/api/RankManager.java
package me.lordsaad.modeoff.api;

import com.google.gson.*;
import com.teamwizardry.librarianlib.features.kotlin.JsonMaker;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.io.*;

/**
 * Created by LordSaad.
 */
public class RankManager {
	public static RankManager INSTANCE = new RankManager();

	public File directory, config;

	private RankManager() {
	}

	@Nullable
	public Rank getPlayerRank(String playerName) {
		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(config));
			if (json.isJsonObject() && json.getAsJsonObject().has("ranks") && json.getAsJsonObject().get("ranks").isJsonArray()) {
				for (JsonElement element : json.getAsJsonObject().getAsJsonArray("ranks")) {
					if (element.isJsonObject()) {
						JsonObject rank = element.getAsJsonObject();
						if (rank.has("players") && rank.get("players").isJsonArray()) {
							JsonArray players = rank.getAsJsonArray("players");
							for (JsonElement player1 : players) {
								if (player1.isJsonPrimitive() && player1.getAsJsonPrimitive().isString()) {
									String name = player1.getAsString();
									if (playerName.equals(name) || playerName.startsWith("Player")) {
										// FOUND RANK
										return new Rank(
												rank.getAsJsonPrimitive("name").getAsString(),
												new Color(rank.getAsJsonPrimitive("color").getAsInt()),
												rank.getAsJsonPrimitive("gm1").getAsBoolean(),
												rank.getAsJsonPrimitive("claimable_plots").getAsInt(),
												rank.getAsJsonPrimitive("manage_others").getAsBoolean());
									}
								}
							}
						}
					}
				}
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		return null;
	}

	public boolean setPlayerRank(String playerName, String rank) {
		boolean changed = false;
		removePlayerRank(playerName);
		try {
			JsonElement json = new JsonParser().parse(new FileReader(config));

			JsonArray currentRanks = json.getAsJsonObject().getAsJsonArray("ranks");

			for (JsonElement element : currentRanks) {
				JsonObject obj = element.getAsJsonObject();

				if (obj.getAsJsonPrimitive("name").getAsString().equals(rank)) {
					obj.getAsJsonArray("players").add(new JsonPrimitive(playerName));
					changed = true;
				}
			}

			FileWriter writer = new FileWriter(config);
			writer.write(JsonMaker.serialize(json));
			writer.flush();
			writer.close();
		} catch (IOException e) {
			e.printStackTrace();
		}

		return changed;
	}

	public void removePlayerRank(String playerName) {
		try {
			JsonElement json = new JsonParser().parse(new FileReader(config));

			boolean removeOld = false;
			String oldRank = null;
			JsonArray oldArray = null;

			if (json.isJsonObject() && json.getAsJsonObject().has("ranks") && json.getAsJsonObject().get("ranks").isJsonArray()) {
				primary:
				for (JsonElement element : json.getAsJsonObject().getAsJsonArray("ranks"))
					if (element.isJsonObject()) {
						JsonObject jsonRank = element.getAsJsonObject();
						if (jsonRank.has("players") && jsonRank.get("players").isJsonArray()) {
							JsonArray players = jsonRank.getAsJsonArray("players");
							for (JsonElement player1 : players)
								if (player1.isJsonPrimitive() && player1.getAsJsonPrimitive().isString())
									if (player1.isJsonPrimitive() && player1.getAsJsonPrimitive().isString()) {
										oldRank = jsonRank.getAsJsonPrimitive("name").getAsString();
										oldArray = players;
										removeOld = true;
										break primary;
									}
						}
					}
			} else return;

			if (removeOld && oldRank != null) {
				JsonArray newArray = new JsonArray();
				for (JsonElement element : oldArray) {
					String name = element.getAsJsonPrimitive().getAsString();
					if (name.equals(playerName)) continue;
					newArray.add(new JsonPrimitive(name));
				}

				JsonArray currentRanks = json.getAsJsonObject().getAsJsonArray("ranks");

				JsonObject newObj = new JsonObject();
				JsonArray newRanks = new JsonArray();
				for (JsonElement element : currentRanks) {
					JsonObject obj = element.getAsJsonObject();

					String name = obj.getAsJsonPrimitive("name").getAsString();
					int color = obj.getAsJsonPrimitive("color").getAsInt();
					boolean gm1 = obj.getAsJsonPrimitive("gm1").getAsBoolean();
					int claimablePlots = obj.getAsJsonPrimitive("claimable_plots").getAsInt();
					boolean manageOthers = obj.getAsJsonPrimitive("manage_others").getAsBoolean();

					if (obj.getAsJsonPrimitive("name").getAsString().equals(oldRank)) {

						JsonObject newRank = new JsonObject();
						newRank.addProperty("name", name);
						newRank.addProperty("color", color);
						newRank.addProperty("gm1", gm1);
						newRank.addProperty("claimable_plots", claimablePlots);
						newRank.addProperty("manage_others", manageOthers);
						newRank.add("players", newArray);

						newRanks.add(newRank);

					} else newRanks.add(element.getAsJsonObject());
				}

				newObj.add("ranks", newRanks);
				FileWriter writer = new FileWriter(config);
				writer.write(JsonMaker.serialize(newObj));
				writer.flush();
				writer.close();
			}
		} catch (IOException e) {
			e.printStackTrace();
		}

	}
}

File: src/main/kotlin/info/modoff/modeoff/api/RankManager.kt
package info.modoff.modeoff.api

import com.google.gson.*
import com.teamwizardry.librarianlib.features.kotlin.serialize

import java.awt.*
import java.io.*

/**
 * Created by LordSaad.
 */
object RankManager {
    var directory: File? = null
    var config: File? = null

    fun getPlayerRank(playerName: String): Rank? {
        try {
            val json = JsonParser().parse(FileReader(config!!))
            if (json is JsonObject) {
                val ranks = json["ranks"]
                if (ranks is JsonArray) {
                    ranks.filterIsInstance<JsonObject>().forEach { rank ->
                        val players = rank["players"]
                        if (players is JsonArray) {
                            players.filterIsInstance<JsonPrimitive>().forEach { name ->
                                if (name.asString == playerName) {
                                    // FOUND RANK
                                    return Rank(
                                        rank.getAsJsonPrimitive("name").asString,
                                        Color(rank.getAsJsonPrimitive("color").asInt),
                                        rank.getAsJsonPrimitive("gm1").asBoolean,
                                        rank.getAsJsonPrimitive("claimable_plots").asInt,
                                        rank.getAsJsonPrimitive("manage_others").asBoolean
                                    )
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        return null
    }

    fun setPlayerRank(playerName: String, rank: String): Boolean {
        var changed = false
        removePlayerRank(playerName)
        try {
            val json = JsonParser().parse(FileReader(config!!))

            val currentRanks = json.asJsonObject.getAsJsonArray("ranks")

            for (element in currentRanks) {
                val obj = element.asJsonObject

                if (obj.getAsJsonPrimitive("name").asString == rank) {
                    obj.getAsJsonArray("players").add(JsonPrimitive(playerName))
                    changed = true
                }
            }

            val writer = FileWriter(config!!)
            writer.write(json.serialize())
            writer.flush()
            writer.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return changed
    }

    fun removePlayerRank(playerName: String) {
        try {
            val json = JsonParser().parse(FileReader(config!!))

            var removeOld = false
            var oldRank: String? = null
            var oldArray: JsonArray? = null

            if (json.isJsonObject && json.asJsonObject.has("ranks") && json.asJsonObject.get("ranks").isJsonArray) {
                primary@ for (element in json.asJsonObject.getAsJsonArray("ranks"))
                    if (element.isJsonObject) {
                        val jsonRank = element.asJsonObject
                        if (jsonRank.has("players") && jsonRank.get("players").isJsonArray) {
                            val players = jsonRank.getAsJsonArray("players")
                            for (player1 in players)
                                if (player1.isJsonPrimitive && player1.asJsonPrimitive.isString)
                                    if (player1.isJsonPrimitive && player1.asJsonPrimitive.isString) {
                                        oldRank = jsonRank.getAsJsonPrimitive("name").asString
                                        oldArray = players
                                        removeOld = true
                                        break@primary
                                    }
                        }
                    }
            } else
                return

            if (removeOld && oldRank != null) {
                val newArray = JsonArray()
                for (element in oldArray!!) {
                    val name = element.asJsonPrimitive.asString
                    if (name == playerName) continue
                    newArray.add(JsonPrimitive(name))
                }

                val currentRanks = json.asJsonObject.getAsJsonArray("ranks")

                val newObj = JsonObject()
                val newRanks = JsonArray()
                for (element in currentRanks) {
                    val obj = element.asJsonObject

                    val name = obj.getAsJsonPrimitive("name").asString
                    val color = obj.getAsJsonPrimitive("color").asInt
                    val gm1 = obj.getAsJsonPrimitive("gm1").asBoolean
                    val claimablePlots = obj.getAsJsonPrimitive("claimable_plots").asInt
                    val manageOthers = obj.getAsJsonPrimitive("manage_others").asBoolean

                    if (obj.getAsJsonPrimitive("name").asString == oldRank) {

                        val newRank = JsonObject()
                        newRank.addProperty("name", name)
                        newRank.addProperty("color", color)
                        newRank.addProperty("gm1", gm1)
                        newRank.addProperty("claimable_plots", claimablePlots)
                        newRank.addProperty("manage_others", manageOthers)
                        newRank.add("players", newArray)

                        newRanks.add(newRank)

                    } else
                        newRanks.add(element.asJsonObject)
                }

                newObj.add("ranks", newRanks)
                val writer = FileWriter(config!!)
                writer.write(newObj.serialize())
                writer.flush()
                writer.close()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }
}
